crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -0,0 +1,929 @@
+            ProviderKind::Replicated => {
+                ReplicatedLogletParams::deserialize_from(value.params.as_bytes())
+                    .map(LogletConfiguration::Replicated)
+                    .map_err(Into::into)
+            }
+        }
+    }
+}
+
+/// Effects used by [`LogsControllerInner`] to signal required actions to the [`LogsController`]. Often those
+/// effects require an asynchronous operation whose result is reported back via an [`Event`].
+enum Effect {
+    /// Write [`Logs`] to the metadata store if the previous version matches
+    WriteLogs {
+        logs: Arc<Logs>,
+        previous_version: Version,
+    },
+    /// Seal the given segment of the given [`LogId`].
+    Seal {
+        log_id: LogId,
+        segment_index: SegmentIndex,
+    },
+}
+
+/// Events that the [`LogsControllerInner`] needs to react to. The events can originate from previous [`Effect`]
+/// or from an external source.
+enum Event {
+    /// Successfully written the [`Logs`] configuration with the given version
+    WriteLogsSucceeded(Version),
+    /// Failed writing `logs`
+    WriteLogsFailed {
+        logs: Arc<Logs>,
+        previous_version: Version,
+    },
+    /// Found a newer [`Logs`] version.
+    NewLogs,
+    /// Sealing of the given log segment succeeded
+    SealSucceeded {
+        log_id: LogId,
+        segment_index: SegmentIndex,
+        seal_lsn: Lsn,
+    },
+    /// Sealing of the given log segment failed
+    SealFailed {
+        log_id: LogId,
+        segment_index: SegmentIndex,
+    },
+}
+
+/// The inner logs controller is responsible for provisioning, monitoring, sealing and reconfiguring
+/// of logs. It does so by reacting to [`Event`], [`Logs`], and [`PartitionTable`] changes. To drive
+/// the internal log state machines forward, it emits [`Effect`].
+///
+/// The underlying idea is that [`Logs`] and [`PartitionTable`] hold the ground truth about the
+/// current state of the logs and which logs to provision. Whenever a newer logs or partition table
+/// is found, the logs controller will reset itself accordingly. Partitions that don't have a
+/// corresponding log in logs, will be provisioned.
+///
+/// Since logs contains the ground truth, any log state change that implies a change in [`Logs`]
+/// (e.g. reconfiguration, initial provisioning) results into a write to the metadata store. Since
+/// there can be multiple writes happening concurrently from different logs controllers, we can only
+/// assume that our [`LogsControllerInner::logs_state`] is valid after we see our write succeed. If
+/// a different logs controller writes first, then we have to reset our internal state. To avoid the
+/// complexity of handling multiple in-flight logs writes, the controller waits until the an
+/// in-flight write completes or a newer log is detected.
+struct LogsControllerInner {
+    // todo make configurable
+    default_provider: ProviderKind,
+
+    logs_state: BTreeMap<LogId, LogState>,
```

Comment:
As a human watching the logs, I find it useful that TreeMap/Set have a predictable order - given the impact not critical, I think this might be a feature right now. 😅